Allow XMLResolver to run without breaking with a restrictive Java security manager

We sometimes use Saxon + XML Resolver on the server side in applications/services which have a restrictive security manager, a security manager which may no allow reading system properties, environmental variables or checking if files exist outside of a certain folder and in such cases the xml resolver may thrown errors like this one for example:

       Caused by: java.security.AccessControlException: access denied ("java.io.FilePermission" "C:\Users\...tomcat\bin\bootstrap.jar" "read")
    at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:485)
    at java.base/java.security.AccessController.checkPermission(AccessController.java:1068)
    at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:416)
    at java.base/java.lang.SecurityManager.checkRead(SecurityManager.java:756)
    at our.custom.CustomSecurityManager.checkRead(CustomSecurityManager.java:120)
    at java.base/java.io.File.exists(File.java:829)
    at org.xmlresolver.XMLResolverConfiguration.findClasspathCatalogFiles(XMLResolverConfiguration.java:1063)
    at org.xmlresolver.XMLResolverConfiguration.showConfig(XMLResolverConfiguration.java:796)

Ideally there would be some kind of setting to instruct the xml resolver that it should not look in system properties, env variables for more XML catalogs. Or if it looks in there, it should attempt to catch AccessControlException errors on its own and thus avoid breaking.

@ndw I tested on our side with the `secmgr` branch, I got rid of all `permission java.lang.RuntimePermission "getenv.XML_CATALOG_XXX";` but still if I remove the java.class.path permission I get:
```
WARN  StageXSLT - [CH][WARN] Cannot instantiate a Configuration
java.lang.RuntimeException: Cannot instantiate a Configuration
        at net.sf.saxon.Configuration.newConfiguration(Configuration.java:277)
        at net.sf.saxon.s9api.Processor.<init>(Processor.java:83)
        at net.sf.saxon.jaxp.SaxonTransformerFactory.<init>(SaxonTransformerFactory.java:59)
        at net.sf.saxon.TransformerFactoryImpl.<init>(TransformerFactoryImpl.java:42)
        ...
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:282)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.security.AccessControlException: access denied ("java.util.PropertyPermission" "java.class.path" "read")
        at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
        at java.base/java.security.AccessController.checkPermission(AccessController.java:897)
        at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:322)
        at java.base/java.lang.SecurityManager.checkPropertyAccess(SecurityManager.java:1066)
        at java.base/java.lang.System.getProperty(System.java:829)
        at org.xmlresolver.XMLResolverConfiguration.findClasspathCatalogFiles(XMLResolverConfiguration.java:1075)
        at org.xmlresolver.XMLResolverConfiguration.showConfig(XMLResolverConfiguration.java:813)
        at org.xmlresolver.XMLResolverConfiguration.loadConfiguration(XMLResolverConfiguration.java:441)
        at org.xmlresolver.XMLResolverConfiguration.<init>(XMLResolverConfiguration.java:288)
        at net.sf.saxon.lib.CatalogResourceResolver.<init>(CatalogResourceResolver.java:46)
        at net.sf.saxon.Configuration.init(Configuration.java:475)
        at net.sf.saxon.Configuration.<init>(Configuration.java:253)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at net.sf.saxon.Configuration.newConfiguration(Configuration.java:271)
        ... 23 common frames omitted
```

--- a/src/main/java/org/xmlresolver/XMLResolverConfiguration.java
+++ b/src/main/java/org/xmlresolver/XMLResolverConfiguration.java
@@ -1071,8 +1071,25 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
             // Starting in Java 9, the class loader is no longer a URLClassLoader, so the .getURLs()
             // trick doesn't work. Instead, we're just going to assume that the java.class.path
             // system property is correct. It seems to be.
-            String sep = System.getProperty("path.separator");
-            String cpath = System.getProperty("java.class.path");
+            String sep;
+            String cpath;
+
+            try {
+                sep = System.getProperty("path.separator");
+            } catch (AccessControlException ex) {
+                // I guess you're not allowed to do this...
+                sep = null;
+                resolverLogger.debug("Access forbidden to environment variable: path.separator");
+            }
+
+            try {
+                cpath = System.getProperty("java.class.path");
+            } catch (AccessControlException ex) {
+                // I guess you're not allowed to do this...
+                cpath = null;
+                resolverLogger.debug("Access forbidden to environment variable: java.class.path");
+            }
+
             if (sep != null && cpath != null) {
                 resolverLogger.log(AbstractLogger.CONFIG, "Searching for catalogs on classpath:");
                 for (String loc : cpath.split(sep)) {

--- a/src/main/java/org/xmlresolver/XMLResolverConfiguration.java
+++ b/src/main/java/org/xmlresolver/XMLResolverConfiguration.java
@@ -21,6 +21,7 @@ import java.lang.reflect.InvocationTargetException;
 import java.net.*;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.security.AccessControlException;
 import java.util.*;
 import java.util.function.Supplier;
 
@@ -339,15 +340,21 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
     }
 
     private String getConfigProperty(String name) {
-        String property = System.getProperty(name);
-        if (property == null) {
-            String env = name
-                    .replaceAll("\\.", "_")
-                    .replaceAll("([a-z])([A-Z])", "$1_$2")
-                    .toUpperCase();
-            property = System.getenv(env);
+        try {
+            String property = System.getProperty(name);
+            if (property == null) {
+                String env = name
+                        .replaceAll("\\.", "_")
+                        .replaceAll("([a-z])([A-Z])", "$1_$2")
+                        .toUpperCase();
+                property = System.getenv(env);
+            }
+            return property;
+        } catch (AccessControlException ex) {
+            // I guess you're not allowed to do this
+            resolverLogger.debug("Access forbidden to system property: " + name);
+            return null;
         }
-        return property;
     }
 
     private void loadConfiguration(List<URL> propertyFiles, List<String> catalogFiles) {
@@ -362,18 +369,28 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
             // to avoid loading the XMLRESOLVER_PROPERTIES environment. This is
             // an edge case more-or-less exclusively for testing.
             if (propfn == null || "".equals(propfn)) {
-                URL propurl = XMLResolverConfiguration.class.getResource("/xmlresolver.properties");
-                if (propurl != null) {
-                    propertyFilesList.add(propurl);
+                try {
+                    URL propurl = XMLResolverConfiguration.class.getResource("/xmlresolver.properties");
+                    if (propurl != null) {
+                        propertyFilesList.add(propurl);
+                    }
+                } catch (AccessControlException ex) {
+                    // I guess you're not allowed to do this
+                    resolverLogger.debug("Access forbidden to class resource");
                 }
             } else {
-                URI baseURI = URIUtils.cwd();
-                for (String fn : propfn.split("\\s*;\\s*")) {
-                    try {
-                        propertyFilesList.add(baseURI.resolve(fn).toURL());
-                    } catch (MalformedURLException ex) {
-                        // nevermind
+                try {
+                    URI baseURI = URIUtils.cwd();
+                    for (String fn : propfn.split("\\s*;\\s*")) {
+                        try {
+                            propertyFilesList.add(baseURI.resolve(fn).toURL());
+                        } catch (MalformedURLException ex) {
+                            // nevermind
+                        }
                     }
+                } catch (AccessControlException ex) {
+                    // I guess you're not allowed to do this
+                    resolverLogger.debug("Access forbidden to cwd");
                 }
             }
         } else {
@@ -390,7 +407,7 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
                     propurl = url;
                     break;
                 }
-            } catch (IOException ex) {
+            } catch (IOException|AccessControlException ex) {
                 // nevermind
             }
         }
@@ -409,7 +426,7 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
         } else {
             catalogs.clear();
             for (String fn : catalogFiles) {
-                if ("".equals(fn.trim())) {
+                if (fn.trim().isEmpty()) {
                     continue;
                 }
                 catalogs.add(fn);
@@ -1060,11 +1077,16 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
                 resolverLogger.log(AbstractLogger.CONFIG, "Searching for catalogs on classpath:");
                 for (String loc : cpath.split(sep)) {
                     File dir = new File(loc);
-                    if (dir.exists() && dir.isDirectory()) {
-                        Path path = Paths.get(loc, "catalog.xml");
-                        if (path.toFile().exists()) {
-                            catalogs.add(path.toString());
+                    try {
+                        if (dir.exists() && dir.isDirectory()) {
+                            Path path = Paths.get(loc, "catalog.xml");
+                            if (path.toFile().exists()) {
+                                catalogs.add(path.toString());
+                            }
                         }
+                    } catch (AccessControlException ex) {
+                        // I guess you're not allowed to do this...
+                        resolverLogger.debug("Access forbidden to file: " + dir);
                     }
                 }
             }
@@ -1075,7 +1097,7 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
                     URL catalog = resources.nextElement();
                     catalogs.add(catalog.toString());
                 }
-            } catch (IOException ex) {
+            } catch (IOException|AccessControlException ex) {
                 // nevermind
             }
 
@@ -1220,10 +1242,19 @@ public class XMLResolverConfiguration implements ResolverConfiguration {
 
     private static class FallbackLogger extends AbstractLogger {
         private final ArrayList<Message> messages = new ArrayList<>();
-        private final String fallbackLogging
-                = System.getProperty("xml.catalog.FallbackLoggerLogLevel") != null
-                ? System.getProperty("xml.catalog.FallbackLoggerLogLevel")
-                : System.getenv("XML_CATALOG_FALLBACK_LOGGER_LOG_LEVEL");
+        private final String fallbackLogging;
+
+        public FallbackLogger() {
+            String logging = null;
+            try {
+                logging = System.getProperty("xml.catalog.FallbackLoggerLogLevel") != null
+                        ? System.getProperty("xml.catalog.FallbackLoggerLogLevel")
+                        : System.getenv("XML_CATALOG_FALLBACK_LOGGER_LOG_LEVEL");
+            } catch (AccessControlException ex) {
+                // I guess you're not allowed to do this
+            }
+            fallbackLogging = logging;
+        }
 
         @Override
         public void log(String cat, String message, Object... params) {

